torch/autograd/gradcheck.py

```diff
@@ -965,17 +965,22 @@ def wrapped_fn(*inputs):
 
     return apply_to_c_outs(fn, torch.real), apply_to_c_outs(fn, torch.imag)
 
-def _real_and_imag_input(fn, complex_inp_indices):
-    # returns new functions that take real inputs instead of complex inputs and compute fn(x + 0 * 1j)
-    # and f(x * 1j).
+def _real_and_imag_input(fn, complex_inp_indices, tupled_inputs):
+    # returns new functions that take real inputs instead of complex inputs as
+    # (x, y) -> fn(x + y * 1j). And it computes: inp -> fn(inp + y * 1j) and inp -> fn(x + inp * 1j).
+    # In each case, the other part is considered constant.
+    # We do not use 0 for the constant here to make sure we always call the user function with a valid input.
```

Comment:
what does this last line mean?

Reply:
The idea is that some function (like sqrt or a few linalg functions) are not defined on the entire space if inputs.
So even though the user gives in input x+iy that is valid, if we replace it with 0+iy, this might not be a valid input anymore.
So we want to preserve the x+iy (even though only one of them is a differentiated value and the other is a "constant".

Reply:
What are examples of functions for which this happens?

The matrix functions we have are (by definition pretty much) meromorphic (e.g. solvers, `inv`, `matrix_exp`). As such, they are defined on the whole C but on a discrete set of points. For matrices, this should be thought of as "they are defined for any matrix but those with certain spectrum". This set is lower-dimensional, so I do not see how this would be a problem.

A similar thing happens for `sqrt` when seen as a function from C -> C. This function is defined (modulo a choice of branch on a ray emanating from zero, for example, the negative numbers) on the whole C. Then, after the choice of branch is made, this function is differentiable (with a relaxed definition of differentiability) on all C but at zero.

A similar thing happens with matrix factorisations, which are just inverse functions but on matrices, similar to the square root.

Reply:
Some trigonometric functions are defined not defined on the axis. So `iy` is not a valid input even though `x + iy` is.
Note that when I say valid input, this means a point where the function is continuously differentiable (so that the finite difference is working fine).

I am not sure if there is an easy way to say that though.

Reply:
After an offline discussion, we agreed that this comment is a bit unfortunate. Removing it will be best. The code below does the right thing indeed.

Reply:
That makes sense. However, not sure why was this causing issue for `torch.chunk`?